### ISSUE TITLE
[Bug] The property 'options.family' must be one of: 0, 4, 6. Received '4'

### ISSUE THREAD

**Issue by nktnet1** [OptimalBits/bull#2704]
## Description

When a REDIS_URL specifies a "family" in the URL, e.g. 
```
redis://@redis:50004/?family=4
```
it is not type-casted to number, resulting in the error
```text
node:dns:179
          validateOneOf(options.family, 'options.family', validFamilies);
          ^

TypeError [ERR_INVALID_ARG_VALUE]: The property 'options.family' must be one of: 0, 4, 6. Received '4'
    at lookup (node:dns:179:11)
    at emitLookup (node:net:1412:5)
    at defaultTriggerAsyncIdScope (node:internal/async_hooks:464:18)
    at lookupAndConnectMultiple (node:net:1411:3)
    at node:net:1357:7
    at defaultTriggerAsyncIdScope (node:internal/async_hooks:464:18)
    at lookupAndConnect (node:net:1356:5)
    at Socket.connect (node:net:1253:5)
    at connect (node:net:236:17)
    at /home/nktnet/temp/bull/node_modules/ioredis/built/connectors/StandaloneConnector.js:54:66 {
  code: 'ERR_INVALID_ARG_VALUE'
}
```

## Minimal, Working Test code to reproduce the issue.

```javascript
const Queue = require('bull');

const myQueue = new Queue('myQueue', 'redis://@redis:50004/?family=4');

myQueue.process(() => {});
```

## Bull version

4.12.2

## Additional information

![1707728023](https://github.com/OptimalBits/bull/assets/86177399/514b73e3-1c54-4b7e-9754-9a642f7f03f7)

- Discovered from https://github.com/nestjs/bull/issues/2001
- My use case is for a URL that requires family=6, e.g.
    ```
    redis://default:c9ee1f12c98411ee8dcf93e49a39cc81@hopper.redis.io:6379/?family=6
    ```

One potential workaround could be to change the `redisOptsFromUrl` function https://github.com/OptimalBits/bull/blob/60fa88f08637f0325639988a3f054880a04ce402/lib/queue.js#L319-L341

to typecast all numbers, e.g.
```javascript
function redisOptsFromUrl(urlString) {
  let redisOpts = {};
  try {
    const redisUrl = url.parse(urlString, true, true);
    redisOpts.port = parseInt(redisUrl.port || '6379', 10);
    redisOpts.host = redisUrl.hostname;
    redisOpts.db = redisUrl.pathname ? redisUrl.pathname.split('/')[1] : 0;
    if (redisUrl.auth) {
      const columnIndex = redisUrl.auth.indexOf(':');
      redisOpts.password = redisUrl.auth.slice(columnIndex + 1);
      if (columnIndex > 0) {
        redisOpts.username = redisUrl.auth.slice(0, columnIndex);
      }
    }

    if (redisUrl.query) {
      for (let key in redisUrl.query) {
        if (!isNaN(redisUrl.query[key])) {
          redisOpts[key] = parseInt(redisUrl.query[key], 10);
        } else {
          redisOpts[key] = redisUrl.query[key];
        }
      }
    }
  } catch (e) {
    throw new Error(e.message);
  }
  return redisOpts;
}
```

```diff
17c17,23
<        redisOpts = { ...redisOpts, ...redisUrl.query };
---
>       for (let key in redisUrl.query) {
>         if (!isNaN(redisUrl.query[key])) {
>           redisOpts[key] = parseInt(redisUrl.query[key], 10);
>         } else {
>           redisOpts[key] = redisUrl.query[key];
>         }
>       }
```
although not too sure about any side effects.

**Comment by manast** [comment]
I am not sure this is an issue in Bull or in IORedis. However you should be able to workaround it by just passing the family options separately:
```ts
const myQueue = new Queue('myQueue', 'redis://@redis:50004/', { redis: { family: 4 } });
```

**Comment by nktnet1** [comment]
Hi @manast,

Thanks for the reply. I use both ioredis and bull and ioredis parses it correctly.

The issue is with this line of code for bull, which parses all query parameters as string:
https://github.com/OptimalBits/bull/blob/60fa88f08637f0325639988a3f054880a04ce402/lib/queue.js#L335

In development, I use `family=4`, but in production it's `family=6`, so the workaround I came up with is in the original comment where I just defined my own `redisOptsFromUrl` function and pass in those parsed options instead :).
